### PR TITLE
CoreDNS Edge plugin rebirth

### PR DESCRIPTION
**Pull request by rfliegerallison**
A complete refactor/re-design of the CoreDNS Edge plugin.

### CODE REVIEW

File: plugin/edge/edge.go [hunk 1]
```diff
@@ -1,106 +1,238 @@
...
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	pluginName              = "edge"
+	defaultSvcReadInterval  = 2 * time.Second
+	defaultSvcPushInterval  = 3 * time.Second
+	defaultExpire           = 10 * time.Second
+	defaultMaxUpstreamFails = 2
+	maxUpstreams            = 15
```
Comment (m-okeefe) [inline plugin/edge/edge.go]: Just curious, why is max upstreams `15`? does this number come from anywhere?

File: plugin/edge/event.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	// Convert the event contents to a service string.
+	evt.Service = generateServiceDNS(e.Object.(*v1.Service))
+
+	return evt, nil
+}
+
+// Generates a services DNS that looks like my-svc.my-namespace.svc.cluster.external
+func generateServiceDNS(svc *v1.Service) string {
+	return fmt.Sprintf("%s.%s.svc.cluster.external", svc.GetName(), svc.GetNamespace())
```
Comment (m-okeefe) [inline plugin/edge/event.go]: @stephenrlouie -- We got a question about this DNS name after our talk, `.external` not kosher?

File: plugin/edge/edge.go (hunk 1 above)
Comment (rfliegerallison) [reply plugin/edge/edge.go]: This was the hardcoded max in the origin `forward` plugin. Not sure why, but it seemed like a reasonable choice. You can always change it if you want.

File: plugin/edge/services.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		}
+		e.watcher = w
+		for {
+			eventChan := e.watcher.ResultChan()
+			for rawEvent := range eventChan {
+
+				// Convert the watch event to our own service event type.
+				event, err := parseEvent(rawEvent)
+				if err != nil {
+					log.Errorf("couldn't read locally running Kubernetes services: %v", err)
```
Comment (m-okeefe) [inline plugin/edge/services.go]: the append string "couldn't read..." on this specific parse error is the same as line 12 ^^, nit, maybe make the error message more specific

File: plugin/plugin.cfg [hunk 4]
```diff
@@ -48,8 +48,7 @@ file:file
 auto:auto
 secondary:secondary
 etcd:etcd
-optikon-central:wwwin-github.cisco.com/edge/optikon-dns/plugin/central
-optikon-edge:wwwin-github.cisco.com/edge/optikon-dns/plugin/edge
+edge:wwwin-github.cisco.com/edge/optikon-dns/plugin/edge
```
Comment (m-okeefe) [inline plugin/plugin.cfg]: change to github.com

File: Dockerfile [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
 RUN go get github.com/coredns/coredns
 RUN go get github.com/opentracing/opentracing-go
+RUN go get github.com/sirupsen/logrus
+RUN go get github.com/mitchellh/hashstructure
+RUN go get k8s.io/client-go/...
+RUN rm -rf /go/src/github.com/coredns/coredns/vendor/github.com/golang/glog
 
 # Mount the central and edge plugins.
-COPY plugin/central /go/src/wwwin-github.cisco.com/edge/optikon-dns/plugin/central
 COPY plugin/edge /go/src/wwwin-github.cisco.com/edge/optikon-dns/plugin/edge
```
Comment (m-okeefe) [inline Dockerfile]: change to github.com

File: plugin/edge/edge.go (hunk 1 above)
Comment (m-okeefe) [reply plugin/edge/edge.go]: np, just curious

**Comment by m-okeefe** [comment]
regarding directory structure, now that you've collapsed edge and central into 1 plugin, maybe instead of `/plugin/edge/....` it's now just `/plugin/....`?

**Comment by rfliegerallison** [comment]
@m-okeefe I think there will likely be additional plugins added soon, so I think it's best to maintain this directory structure for now.

File: plugin/edge/event.go (hunk 2 above)
Comment (stephenrlouie) [reply plugin/edge/event.go]: I just think `.external` is super expensive. Lets just do `.optikon` for now

File: Dockerfile [hunk 6]
```diff
@@ -1,13 +1,16 @@
 # Start with a Golang-enabled base image.
 FROM golang:1.10.0 as builder
 
-# Fetch the CoreDNS repo.
+# Fetch all dependencies.
 RUN go get github.com/coredns/coredns
 RUN go get github.com/opentracing/opentracing-go
+RUN go get github.com/sirupsen/logrus
```
Comment (stephenrlouie) [inline Dockerfile]: remind me why these are not vendored?

File: Makefile [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # Makefile for Optikon DNS.
 
 IMAGE ?= intelligentedgeadmin/optikon-dns
```
Comment (stephenrlouie) [inline Makefile]: update to reflect new dockerhub account

File: README.md [hunk 8]
```diff
@@ -1,3 +1,3 @@
 # Optikon DNS
 
-Custom DNS for Optikon, Using Homemade CoreDNS Plugins
+Custom DNS for Optikon, Using Homemade CoreDNS Plugins.
```
Comment (stephenrlouie) [inline README.md]: Lets buff this up a little, we can throw one of the diagrams we used on the presentation.

File: manifests/kube-dns-depl.yaml [hunk 9]
```diff
@@ -33,7 +33,7 @@ spec:
       - args:
         - -conf
         - /etc/coredns/Corefile
-        image: intelligentedgeadmin/optikon-dns:1.0.0
+        image: intelligentedgeadmin/optikon-dns:2.1.0
```
Comment (stephenrlouie) [inline manifests/kube-dns-depl.yaml]: reflect new dockerhub

File: plugin/edge/concurrent_service_table.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		edgeSites.Add(meta)
+	} else {
+		newSet := NewSet()
+		newSet.Add(meta)
+		cst.table[serviceName] = newSet
+	}
+
+	// Log the new table.
+	if svcDebugMode {
+		log.Infof("Updated table: %+v", cst.table)
```
Comment (m-okeefe) [inline plugin/edge/concurrent_service_table.go]: This log message is pretty noisy on `central`. Just thinking about scale here -- spamming the entire table to stdout best?

File: plugin/edge/loc_record.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	// Use regex to parse out the parts of the string.
+	parts := locRe.FindStringSubmatch(l)
+	if parts == nil || len(parts) != 13 {
+		log.Errorf("LOC string didn't match regex: recieved parts %+v", parts)
+		return 0.0, 0.0, errInvalidLOC
+	}
+
+	// Log the parsed parts.
+	if dnsDebugMode {
+		log.Infof("parsed latitude_degrees=%s, latitude_minutes=%s, latitude_seconds=%s, latitude_direction=%s, longitude_degrees=%s, longitude_minutes=%s, longitude_seconds=%s, longitude_direction=%s, altitude=%s, size=%s, horizontal_precision=%s, vertical_precision=%s", parts[1], parts[2], parts[3], parts[4], parts[5], parts[6], parts[7], parts[8], parts[9], parts[10], parts[11], parts[12])
```
Comment (m-okeefe) [inline plugin/edge/loc_record.go]: can you make this a debug level log?

Comment (m-okeefe) [reply plugin/edge/loc_record.go]: and maybe any of the location conversion logs, make those debug too?

**Comment by stephenrlouie** [comment]
extension `.optikon` instead of `.external`

File: plugin/edge/concurrent_service_table.go (hunk 10 above)
Comment (rfliegerallison) [reply plugin/edge/concurrent_service_table.go]: I suppose we could introduce another level of debug logging. I think it's easier to see the current state of the table within a couple lines and use Cmd+F to find what you're looking for, as opposed to scrolling up and trying to look for each individual event that came through over potentially a very long time.

File: Dockerfile (hunk 6 above)
Comment (rfliegerallison) [reply Dockerfile]: The entire coredns repo is the only piece that's annoying to vendor

File: plugin/edge/concurrent_service_table.go (hunk 10 above)
Comment (stephenrlouie) [reply plugin/edge/concurrent_service_table.go]: I saw you imported a logger, that should control log levels without the `if`

File: plugin/edge/geo.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
+package edge
```
Comment (stephenrlouie) [inline plugin/edge/geo.go]: why not a lib to do this.

Comment (stephenrlouie) [inline plugin/edge/geo.go]: good for now. I'm sure there is a lib we could use to do the same / be optimized

File: plugin/edge/proxy.go [hunk 13]
```diff
@@ -40,12 +94,11 @@ func NewProxy(addr string, tlsConfig *tls.Config) *Proxy {
 func dnsClient(tlsConfig *tls.Config) *dns.Client {
 	c := new(dns.Client)
 	c.Net = "udp"
-	// TODO(miek): this should be half of hcDuration?
+	// TODO(miek): this should be half of healthCheckDuration?
```
Comment (stephenrlouie) [inline plugin/edge/proxy.go]: ?

File: plugin/edge/proxy.go [hunk 14]
```diff
@@ -63,30 +116,65 @@ func (p *Proxy) Dial(proto string) (*dns.Conn, error) { return p.transport.Dial(
...
+	}
+	req, err := http.NewRequest("POST", p.pushAddr, bytes.NewBuffer(jsn))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	go func() {
+		for {
+			resp, err := client.Do(req)
```
Comment (stephenrlouie) [inline plugin/edge/proxy.go]: `defer resp.Body.Close()`

Comment (stephenrlouie) [inline plugin/edge/proxy.go]: if you wrap this in a `func` you can `defer resp.Body.Close()`. I was going to recommend putting the entire go routine instantiation in a loop. but I fear run away go routines. We should also properly track this go routine. if it dies or exits, we should restart.

File: plugin/edge/services.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
+package edge
+
+import (
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Starts the process of reading Kubernetes services using a list watcher.
+func (e *Edge) startReadingServices() {
+	go func() {
```
Comment (stephenrlouie) [inline plugin/edge/services.go]: Do we have to notify on error? We should also ensure a clean exit.

File: plugin/edge/edge.go [hunk 16]
```diff
@@ -112,151 +264,130 @@ func (oe *OptikonEdge) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
...
 
-		return 0, nil
+// Determines the IP address of the edge site closest to the given Point.
+func findClosestToPoint(edgeSiteSet Set, p Point) net.IP {
+	var closest net.IP
+	var minDist float64
+	first := true
+	for _, val := range edgeSiteSet {
+		edgeSite := val.(Site)
+		dist := p.GreatCircleDistance(edgeSite.GeoCoords)
```
Comment (stephenrlouie) [inline plugin/edge/edge.go]: With lots of sites, this is expensive. We don't have to take this optimization on right now, but lets leave an issue for it.

File: README.md (hunk 8 above)
Comment (stephenrlouie) [reply README.md]: I'll take a shot at it and submit a PR for this branch. Please correct whatever I submit and feel free to add more detail

**Pull request merged**